### PR TITLE
pkg/utils/etcdutil: add opt-in watcher reload after compaction

### LINKED ISSUE #11032: LoopWatcher can miss changes after etcd compaction
### What did you do?

Disconnect a LoopWatcher, update its watched key, advance the etcd
revision, compact past the missed update, and then reconnect the watcher.

This was found while investigating:
https://github.com/tidbcloud/pd-cse/issues/559

### What did you expect to see?

After detecting a compacted watch revision, LoopWatcher should reload a
consistent snapshot and continue watching from the snapshot revision plus
one.

### What did you see instead?

LoopWatcher changes its watch revision to the compact revision without
reloading. Changes in the compacted interval are not replayed, so the
consumer cache can remain stale indefinitely.

Paginated loading can also read different revisions, and an earlier
callback error can be overwritten by a later successful callback.

### What version of PD are you using (`pd-server -V`)?

master at 39b622049

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

Issue Number: ref #11032

When an etcd watch revision has been compacted, LoopWatcher normally resumes
from the compact revision without rebuilding current state. Bounded consumers
that own a complete cache need an explicit full reload so they can recover
missed updates and deleted keys.

High-cardinality and streaming consumers cannot safely publish a partial
reload, so this PR keeps their existing pagination and callback semantics.
Their staged snapshot support is handled by follow-up #11118.

### What is changed and how does it work?

- Add opt-in, revision-consistent reload after compaction.
- Preserve the first callback error and advance from `R+1` only after a
  successful opt-in reload.
- Reconcile deleted keys only for bounded caches that explicitly enable it.
- Keep default watchers on their legacy per-page loading path without an
  extra key index, etcd RPC, or per-key reconciliation branch.
- Retry compaction resync locally and do not recreate a watch until reload
  succeeds.
- Enable the opt-in behavior only for bounded singleton, registry, store, and
  scheduling-config watchers.
- Add regression coverage for default compatibility, compaction recovery,
  snapshot consistency, callback errors, deletion reconciliation, and retry
  backoff.

```commit-message
Add opt-in LoopWatcher reload after compaction for bounded consumers while
preserving the default watcher loading contract.
```

### Check List

Tests

- Unit test
- Integration test
- Race test
- Static check
- Large-cardinality benchmark

### Release note

```release-note
Fix stale bounded LoopWatcher caches after etcd watch compaction without changing high-cardinality watcher behavior.
```

### CODE REVIEW

File: pkg/utils/etcdutil/etcdutil.go [hunk 1]
```diff
@@ -768,3 +879,11 @@ func (lw *LoopWatcher) SetLoadRetryTimes(times int) {
 func (lw *LoopWatcher) SetLoadBatchSize(size int64) {
 	lw.loadBatchSize = size
 }
+
+// SetReconcileDeletedKeys enables deletion reconciliation for full loads.
+// It must be called before StartWatchLoop. Since this keeps one entry per
+// watched key, callers should only enable it for prefixes with bounded key cardinality.
+func (lw *LoopWatcher) SetReconcileDeletedKeys() {
```
Comment (lhy1024) [inline pkg/utils/etcdutil/etcdutil.go:979]: Should we also opt the ResourceManager metadata watcher into this mechanism in this PR? It maintains cached resource-group settings and service limits and has a non-no-op delete callback, but its factory exposes only StartWatchLoop and WaitLoad, so it currently cannot call this setter. I verified the generic behavior with a focused unit test: a reload without this opt-in retains a deleted key, while TestWatcherReconcilesDeletedKeysAfterCompaction passes with it enabled. If this watcher is intentionally out of scope, could we document or narrow the stated coverage?

File: pkg/mcs/scheduling/server/rule/watcher.go [hunk 2]
```diff
@@ -273,6 +282,7 @@ func (rw *Watcher) initializeRegionLabelWatcher() error {
 		postEventsFn,
 		true, /* withPrefix */
 	)
+	rw.labelWatcher.SetReconcileDeletedKeys()
```
Comment (lhy1024) [inline pkg/mcs/scheduling/server/rule/watcher.go]: Should rw.ruleWatcher be opted in here as well? This change enables reconciliation for labelWatcher, but the placement-rule watcher also maintains cached Rule and RuleGroup state and has a delete callback. Its previously applied rules are persisted in local RuleStorage by TryCommitPatchLocked, so it has the state needed for a synthetic delete during reload. Without this opt-in, a compaction reload reapplies surviving rules but cannot remove a deleted one. I verified the generic behavior with a focused unit test. Is this path intentionally out of scope for this PR, or should we cover it with the same opt-in and a delete-plus-compaction regression test?

File: pkg/utils/etcdutil/etcdutil.go [hunk 3]
```diff
@@ -621,36 +654,78 @@ func (lw *LoopWatcher) watch(ctx context.Context, revision int64) (nextRevision
...
+		if postErr := lw.postEventsFn([]*clientv3.Event{}); postErr != nil {
 			log.Error("run post event failed in watch loop", zap.String("name", lw.name),
-				zap.String("key", lw.key), zap.Error(err))
+				zap.String("key", lw.key), zap.Error(postErr))
+			if err == nil {
+				err = postErr
+			}
+		}
+		if loadCompleted && err == nil && lw.reconcileDeletedKeys {
+			lw.loadedKeys = snapshotKeys
```
Comment (lhy1024) [inline pkg/utils/etcdutil/etcdutil.go:753]: This all-or-nothing replacement loses partial reconciliation progress. `reconcileLoadedKeys` can successfully apply deletion A, then fail deletion B, while a successful `postEventsFn` still commits A. A remains in `loadedKeys`, so the retry invokes A delete again. That is unsafe for existing non-idempotent callbacks, including the rule watcher after its first commit removes the local RuleStorage entry, and the TSO watcher when the service-registry map entry is gone. A transient B failure can therefore leave compaction reload retrying forever.

I reproduced this on `03ed7ea`: a two-key focused test with one successful delete and one one-time delete failure fails the second reload with `key was already deleted`; the same test passes on `492c88c`. Please record successful operations at the same successful post-callback commit boundary, or make the callback contract explicitly idempotent, and add a regression test such as:

```go
func (suite *loopWatcherTestSuite) TestWatcherRetriesPartialReconciliationDelete() {
	re := suite.Require()
	const prefix = "TestWatcherRetriesPartialReconciliationDelete/"
	firstKey, secondKey := prefix+"first", prefix+"second"
	suite.put(re, firstKey, "first")
	suite.put(re, secondKey, "second")

	cache := make(map[string]string)
	transientErr := errors.New("transient delete error")
	alreadyDeletedErr := errors.New("key was already deleted")
	failSecondDelete := true
	watcher := NewLoopWatcher(
		suite.ctx, &suite.wg, suite.client, "test", prefix,
		func([]*clientv3.Event) error { return nil },
		func(kv *mvccpb.KeyValue) error {
			cache[string(kv.Key)] = string(kv.Value)
			return nil
		},
		func(kv *mvccpb.KeyValue) error {
			key := string(kv.Key)
			if key == secondKey && failSecondDelete {
				failSecondDelete = false
				return transientErr
			}
			if _, ok := cache[key]; !ok {
				return alreadyDeletedErr
			}
			delete(cache, key)
			return nil
		},
		func([]*clientv3.Event) error { return nil },
		true,
	)
	watcher.SetReconcileDeletedKeys()

	_, err := watcher.load(suite.ctx)
	re.NoError(err)
	_, err = suite.client.Delete(suite.ctx, firstKey)
	re.NoError(err)
	_, err = suite.client.Delete(suite.ctx, secondKey)
	re.NoError(err)

	_, err = watcher.load(suite.ctx)
	re.ErrorIs(err, transientErr)
	_, err = watcher.load(suite.ctx)
	re.NoError(err)
	re.Empty(cache)
}
```

File: pkg/mcs/scheduling/server/rule/watcher_test.go [hunk 4]
```diff
@@ -76,14 +158,15 @@ func runWatcherLoadLabelRule(ctx context.Context, re *require.Assertions, client
 		etcdClient:            client,
 		ruleStorage:           storage,
 		regionLabeler:         labelerManager,
+		checkerController:     checkerController,
```
Comment (lhy1024) [inline pkg/mcs/scheduling/server/rule/watcher_test.go]: Minor: could this helper close `rw` after the assertion instead of only canceling its context? `Watcher.Close()` waits for the label watcher and clears the injected checker controller suspect ranges. `BenchmarkLoadLargeRules` now reuses one controller across iterations, so without that cleanup later samples inherit the previous 16K ranges and can overlap with the previous watcher teardown. A `defer rw.Close()` after successful initialization would make each iteration independent.

File: pkg/mcs/resourcemanager/server/manager.go [hunk 5]
```diff
@@ -97,7 +97,16 @@ type Manager struct {
 	enableMetadataWatcher bool
 	controllerConfig      *ControllerConfig
 	krgms                 map[uint32]*keyspaceResourceGroupManager
-	storage               interface {
+	// metadataSnapshotGeneration is incremented for each full metadata load.
+	// The active generation is attached to concurrent API writes so the
+	// successful reconciliation cannot remove them.
+	metadataSnapshotGeneration       uint64
+	activeMetadataSnapshotGeneration uint64
+	// metadataSnapshotMu closes the boundary race between starting a snapshot
```
Comment (lhy1024) [inline pkg/mcs/resourcemanager/server/manager.go]: **[P1] Preserve API writes that overlap a fixed snapshot**

`metadataSnapshotMu` is released when `beginMetadataSnapshot` returns, while `LoopWatcher.load` obtains its first etcd response and dispatches the snapshot `putFn`s later. If etcd has already fixed snapshot revision R but an API write completes before its old KV is dispatched, the API write and the old snapshot receive the same generation.

The old snapshot can then overwrite the newer cached value and mark it as observed; a deleted group can likewise be recreated. `finishMetadataSnapshot` cannot distinguish either case. The watch event at R+1 repairs it only after the full reload completes, which can leave a long stale quota/resource-group window for a large reload.

Could we establish ordering at snapshot-revision acquisition, or keep a separate API-write/tombstone version so a snapshot callback cannot clobber a later API mutation?

Suggested regression test for `metadata_watcher_test.go` (this fails on `05fde590` with priority `5` instead of `6`):
```go
func TestMetadataSnapshotDoesNotOverwriteOverlappingAPIWrite(t *testing.T) {
    re := require.New(t)
    m := newMetadataWatcherTestManager(storage.NewStorageWithMemoryBackend())

    snapshotGroup := newMetadataWatcherResourceGroup("api_group", 5, 100, 200)
    re.NoError(m.AddResourceGroup(snapshotGroup))
    re.NoError(m.SetKeyspaceServiceLimit(10, 100))

    generation := m.beginMetadataSnapshot()
    apiGroup := newMetadataWatcherResourceGroup("api_group", 6, 100, 200)
    re.NoError(m.ModifyResourceGroup(apiGroup))
    re.NoError(m.SetKeyspaceServiceLimit(10, 200))

    // Model a KV from the snapshot revision before the API writes.
    re.NoError(m.handleMetadataWatchPutWithGeneration(
        "resource_group/keyspace/settings/10/api_group",
        mustMarshalResourceGroup(t, snapshotGroup),
        generation,
    ))
    re.NoError(m.handleMetadataWatchPutWithGeneration(
        "resource_group/keyspace/service_limits/10",
        "100",
        generation,
    ))
    re.NoError(m.finishMetadataSnapshot(generation, nil))

    krgm := m.getKeyspaceResourceGroupManager(10)
    re.Equal(uint32(6), krgm.getResourceGroup("api_group", false).Priority)
    re.InDelta(200, m.GetKeyspaceServiceLimiter(10).ServiceLimit, 0.00001)
}
```

File: pkg/mcs/resourcemanager/server/metadata_watcher.go [hunk 6]
```diff
@@ -185,19 +201,38 @@ func (m *Manager) initializeMetadataWatcher(ctx context.Context) error {
...
+
+func (m *Manager) handleMetadataWatchPutWithGeneration(
+	key, rawValue string,
+	metadataSnapshotGeneration uint64,
+) error {
 	target, ok := parseResourceGroupWatchPath(key)
 	if !ok {
 		return nil
 	}
+	if metadataSnapshotGeneration != 0 {
```
Comment (lhy1024) [inline pkg/mcs/resourcemanager/server/metadata_watcher.go]: **[P1] Do not replay an old runtime-state checkpoint into a live manager**

The new compaction path performs a full reload of the entire `resource_group/` prefix, including `resource_group/.../states/...`. These state callbacks are not idempotent: `SetStatesIntoResourceGroup` replaces the live token-bucket state, while it applies `RUConsumption` through the additive `UpdateRUConsumption` method. Runtime RequestRU and consumption mutations are also not recorded in `metadataSnapshotMutations`, so this guard does not protect them.

Consequently, a compaction reload can roll current tokens back to an older persisted checkpoint and add the same persisted consumption again. Unlike the metadata-write race, an in-memory state change that has not yet been persisted may have no R+1 watch event to restore it. The non-idempotent callback predates this PR, but invoking it during live compaction recovery is introduced by the new reload behavior.

Could we define a state-specific reload policy that cannot rewind or double-count live state? A focused regression test below fails on this head: tokens change from 654 back to 321; placing the RRU assertion first also reports 22 instead of 11.

```go
func TestMetadataSnapshotReloadDoesNotReapplyRuntimeState(t *testing.T) {
    re := require.New(t)
    m := newMetadataWatcherTestManager(storage.NewStorageWithMemoryBackend())
    group := newMetadataWatcherResourceGroup("group", 5, 100, 200)
    settingsKey := "resource_group/keyspace/settings/10/group"
    statesKey := "resource_group/keyspace/states/10/group"
    re.NoError(m.handleMetadataWatchPut(settingsKey, mustMarshalResourceGroup(t, group)))

    checkpointTime := time.Now().Add(-time.Minute)
    checkpoint := &GroupStates{
        RU: &GroupTokenBucketState{Tokens: 321, LastUpdate: &checkpointTime},
        RUConsumption: &rmpb.Consumption{RRU: 11},
    }
    rawCheckpoint, err := json.Marshal(checkpoint)
    re.NoError(err)
    re.NoError(m.handleMetadataWatchPut(statesKey, string(rawCheckpoint)))

    generation := m.beginMetadataSnapshot()
    // Settings sort before states and mark the group as observed by this snapshot.
    re.NoError(m.handleMetadataWatchPutWithGeneration(
        settingsKey, mustMarshalResourceGroup(t, group), generation))
    currentTime := time.Now()
    m.getKeyspaceResourceGroupManager(10).getMutableResourceGroup("group").
        SetStatesIntoResourceGroup(&GroupStates{
            RU: &GroupTokenBucketState{Tokens: 654, LastUpdate: &currentTime},
        })
    re.NoError(m.handleMetadataWatchPutWithGeneration(
        statesKey, string(rawCheckpoint), generation))
    re.NoError(m.finishMetadataSnapshot(generation, nil))

    got := m.getKeyspaceResourceGroupManager(10).getResourceGroup("group", true)
    re.InDelta(654, got.RUSettings.RU.Tokens, 0.001) // actual: 321
    re.InDelta(11, got.RUConsumption.RRU, 0.001)    // actual: 22
}
```

File: pkg/mcs/resourcemanager/server/manager.go [hunk 7]
```diff
@@ -296,6 +312,77 @@ func (m *Manager) getKeyspaceResourceGroupManager(keyspaceID uint32) *keyspaceRe
...
+	if m.metadataSnapshotGeneration == 0 {
+		m.metadataSnapshotGeneration++
+	}
+	m.activeMetadataSnapshotGeneration = m.metadataSnapshotGeneration
+	generation := m.activeMetadataSnapshotGeneration
+	m.Unlock()
+	return generation
+}
+
+func (m *Manager) withMetadataAPIWrite(key string, fn func(uint64) error) error {
```
Comment (lhy1024) [inline pkg/mcs/resourcemanager/server/manager.go]: **[P1] Include controller-config writers in the snapshot mutation ordering**

`SetKeyspaceRUVersion` and `UpdateControllerConfigItem` both mutate and persist `keypath.ControllerConfigPath()`, but neither calls `withMetadataAPIWrite`. If etcd has fixed an older snapshot revision and either API completes before that snapshot KV is dispatched, `metadataSnapshotMutations` has no entry for the controller key and `applyControllerConfigFromRaw` overwrites the successfully written configuration.

The R+1 watch event can repair the cache only after the full reload completes, so RU cost or version policy may remain stale throughout a large reload. This live API-vs-snapshot race becomes reachable because this PR adds compaction-triggered full reloads.

Could both controller-config writers participate in the same ordering with `keypath.ControllerConfigPath()`? The focused regression test below fails on this head with `ReadBaseCost == 1` after the API successfully set it to 2.

```go
func TestMetadataSnapshotDoesNotOverwriteControllerConfigAPIWrite(t *testing.T) {
    re := require.New(t)
    m := newMetadataWatcherTestManager(storage.NewStorageWithMemoryBackend())
    m.controllerConfig.RequestUnit.ReadBaseCost = 1
    staleSnapshot, err := json.Marshal(m.controllerConfig)
    re.NoError(err)

    generation := m.beginMetadataSnapshot()
    re.NoError(m.UpdateControllerConfigItem("request-unit.read-base-cost", 2.0))
    re.NoError(m.handleMetadataWatchPutWithGeneration(
        keypath.ControllerConfigPath(),
        string(staleSnapshot),
        generation,
    ))
    re.NoError(m.finishMetadataSnapshot(generation, nil))

    re.InDelta(2.0, m.GetControllerConfig().RequestUnit.ReadBaseCost, 0.00001)
    // Actual value on this head: 1.
}
```

File: pkg/mcs/resourcemanager/server/manager.go [hunk 8]
```diff
@@ -296,6 +318,84 @@ func (m *Manager) getKeyspaceResourceGroupManager(keyspaceID uint32) *keyspaceRe
...
+func (m *Manager) finishMetadataSnapshot(generation uint64, loadErr error) error {
+	if generation == 0 {
+		return nil
+	}
+	// Wait for every API write that observed this generation before reconciling.
+	// Holding the write lock through the scan also prevents a successful API
+	// mutation from racing with deletion of an entry missing from the snapshot.
+	m.metadataSnapshotMu.Lock()
+	defer m.metadataSnapshotMu.Unlock()
+	if loadErr == nil {
```
Comment (lhy1024) [inline pkg/mcs/resourcemanager/server/manager.go]: [P1] Please prevent a snapshot completion from a stale watcher lifecycle from modifying the newly initialized cache.

On RM step-down, `campaignLeader` only cancels the old primary context; it does not wait for the old watcher in `Manager.wg` to exit before the same process can enter `Init` again. In addition, `EtcdKVGet` uses the client context, so an in-flight final-page load can still complete successfully after the primary context has been canceled.

The new `initializeMetadataWatcher` may therefore reset `krgms` and finish its bootstrap before the old watcher calls this function with its old generation. This currently reconciles the new cache with the stale generation before checking whether that generation is still active, deleting resource groups and service limits loaded by the new lifecycle. If the stale completion runs between the reset and the new snapshot start, it can also set `metadataSnapshotInitialized` again, causing the new bootstrap to be classified as a reload and skip persisted states.

This can leave a serving RM without resource groups, service limits, or runtime checkpoints until another full reload occurs. Could we either wait for the previous manager goroutines before reinitializing, or associate snapshots with a watcher lifecycle epoch and ignore callbacks and completions from stale epochs?

The following deterministic test reproduces the first failure mode. It fails at the final `NotNil` because the stale watcher completion removes the group loaded by the new lifecycle:

```go
func TestStaleWatcherDoesNotReconcileNewLifecycle(t *testing.T) {
	re := require.New(t)
	m := newMetadataWatcherTestManager(storage.NewStorageWithMemoryBackend())
	m.srv = &testBasicServer{}
	group := newMetadataWatcherResourceGroup("group", 5, 100, 200)

	var oldWatcher *fakeMetadataLoopWatcher
	factoryCalls := 0
	withMetadataLoopWatcherFactory(t, func(
		_ context.Context,
		_ *sync.WaitGroup,
		_ *clientv3.Client,
		_, _ string,
		_ func([]*clientv3.Event) error,
		putFn, _ func(*mvccpb.KeyValue) error,
		_ func([]*clientv3.Event) error,
		_ bool,
	) metadataLoopWatcher {
		factoryCalls++
		watcher := &fakeMetadataLoopWatcher{waitLoadFn: func() error {
			return putFn(&mvccpb.KeyValue{
				Key:   []byte("resource_group/keyspace/settings/10/group"),
				Value: []byte(mustMarshalResourceGroup(t, group)),
			})
		}}
		if factoryCalls == 1 {
			oldWatcher = watcher
		}
		return watcher
	})

	oldCtx, cancelOld := context.WithCancel(context.Background())
	re.NoError(m.initializeMetadataWatcher(oldCtx))
	oldWatcher.preLoadFn()
	cancelOld()

	re.NoError(m.initializeMetadataWatcher(context.Background()))
	krgm := m.getKeyspaceResourceGroupManager(10)
	re.NotNil(krgm.getResourceGroup("group", false))

	re.NoError(oldWatcher.postLoadFn(nil))
	re.NotNil(krgm.getResourceGroup("group", false))
}
```

File: pkg/mcs/resourcemanager/server/manager_test.go [hunk 9]
```diff
@@ -162,6 +178,28 @@ func withMetadataLoopWatcherFactory(t *testing.T, factory testMetadataLoopWatche
...
+		m.wg.Add(1)
+		go func() {
+			defer m.wg.Done()
+			<-oldCtx.Done()
+			close(oldExited)
+		}()
+
+		re.NoError(m.Init(context.Background()))
+		defer m.close()
+		select {
```
Comment (lhy1024) [inline pkg/mcs/resourcemanager/server/manager_test.go]: Non-blocking suggestion: could we make this assertion deterministic by controlling when the previous lifecycle goroutine is allowed to exit? The goroutine currently closes `oldExited` immediately after observing cancellation, so even without `m.wg.Wait()` it may happen to run before `Init` returns and let the test pass depending on scheduling. I temporarily removed the wait and observed mixed outcomes across repeated runs, including iterations that passed.

For example, the old goroutine could acknowledge cancellation and then block on a release channel; run the second `Init` asynchronously, verify that it has not returned, release the old goroutine, and finally verify that `Init` completes. That would specifically prove the wait ordering rather than only checking that cancellation eventually scheduled the goroutine.

**Review by lhy1024 — APPROVED** [review]

**Comment by rleungx** [comment]
Split complete:
- #11033 now contains LoopWatcher consistency/retry semantics plus
  bounded-cache deletion reconciliation.
- Resource Manager generation reconciliation and atomic region-label
  snapshot publication moved to
  rleungx:large-watcher-snapshot-reload.

I’ll open the follow-up against master after #11033 merges; opening it now
would make GitHub show both PRs’ changes because the follow-up depends on
#11033.

**Comment by rleungx** [comment]
@bufferflies @JmPotato PTAL

File: pkg/mcs/scheduling/server/config/watcher.go [hunk 10]
```diff
@@ -234,6 +236,7 @@ func (cw *Watcher) initializeSchedulerConfigWatcher() error {
 		func([]*clientv3.Event) error { return nil },
 		true, /* withPrefix */
 	)
+	cw.schedulerConfigWatcher.SetReconcileDeletedKeys()
```
Comment (lhy1024) [inline pkg/mcs/scheduling/server/config/watcher.go:239]: **[P2] Make the scheduler-config callback failure-atomic before relying on reconciliation**

The SaveSchedulerConfig-before-ReloadSchedulerConfig behavior predates this PR, so this is not a newly introduced storage mutation. The issue is that this PR opts the callback into SetReconcileDeletedKeys, whose bookkeeping only records keys whose callbacks return nil.

If the save succeeds but reload returns an error (for example, invalid scheduler configuration), the in-memory storage has already been changed but the key is not added to loadedKeys. If the key is then deleted during the compaction gap, the next reload sees an empty snapshot, has no tracked key to reconcile, and returns success while the deleted scheduler config remains in memory until restart.

Could we roll back SaveSchedulerConfig when ReloadSchedulerConfig fails, or otherwise make the callback's side effects failure-atomic before relying on this reconciliation path? This is a feature-completeness issue in the new recovery path rather than a new regression in the existing watcher.

The following focused regression test fails on this head at the final re.Empty(cache) assertion:

```go
func (suite *loopWatcherTestSuite) TestWatcherReconcilesFailedPutSideEffects() {
    re := suite.Require()
    const prefix = "TestWatcherReconcilesFailedPutSideEffects/"
    key := prefix + "new"
    cache := make(map[string]string)
    callbackErr := errors.New("callback failed after applying the value")
    watcher := NewLoopWatcher(
        suite.ctx,
        &suite.wg,
        suite.client,
        "test",
        prefix,
        func([]*clientv3.Event) error { return nil },
        func(kv *mvccpb.KeyValue) error {
            cache[string(kv.Key)] = string(kv.Value)
            return callbackErr
        },
        func(kv *mvccpb.KeyValue) error {
            delete(cache, string(kv.Key))
            return nil
        },
        func([]*clientv3.Event) error { return nil },
        true,
    )
    watcher.SetReconcileDeletedKeys()

    _, err := watcher.load(suite.ctx)
    re.NoError(err)
    _, err = suite.client.Put(suite.ctx, key, "value")
    re.NoError(err)
    _, err = watcher.load(suite.ctx)
    re.ErrorIs(err, callbackErr)
    re.Equal(map[string]string{key: "value"}, cache)

    _, err = suite.client.Delete(suite.ctx, key)
    re.NoError(err)
    _, err = watcher.load(suite.ctx)
    re.NoError(err)
    re.Empty(cache)
}
```

**Review by bufferflies — APPROVED** [review]

File: pkg/mcs/scheduling/server/config/watcher.go (hunk 10 above)
Comment (lhy1024) [reply pkg/mcs/scheduling/server/config/watcher.go:239]: Correction after tracing the scheduler-config lifecycle more carefully: I no longer think this is an actionable issue in this PR.

The failing test above assumes a first-seen, untracked key whose put callback mutates state and then returns an error. That premise does not match the normal scheduler-config path:

- During startup, the config watcher completes its initial load before the schedulers controller is installed, so existing keys return success and enter `loadedKeys`.
- For a new key at runtime, `ErrSchedulerNotFound` is explicitly converted to success when the scheduler has not been created yet, so the key is still tracked.
- If the scheduler already exists because the global-config watcher was processed first, the scheduler-config value comes from the scheduler creation/update path. The creation path persists the scheduler-encoded config before the global scheduler list, and normal update handlers parse and validate the value before persisting it.

More importantly, a failed update of an already tracked key does not remove the existing entry from `loadedKeys`. I verified this with a focused counter-test using the sequence: successful initial load, update callback mutates state and returns an error, etcd key is deleted, then a full reload. The key remains tracked after the failed update, `deleteFn` is invoked during reconciliation, and the cache is empty at the end.

Therefore, the generic test in the original comment demonstrates a possible `LoopWatcher` callback contract edge case, but it does not establish a production regression for this scheduler-config watcher. A malformed value written directly to etcd could manufacture the premise, but that is not enough to require rollback logic here or to block this PR. Please treat my original P2 as withdrawn.

**Review by lhy1024 — APPROVED** [review]
